core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -443,41 +443,32 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
     val maxBytes = Math.min(fetchMaxBytes, fetchInfo.maxBytes)
     val rlsMetadata = getRemoteLogSegmentMetadata(tp, offset)
 
-    var startPos = lookupPositionForOffset(rlsMetadata, offset)
+    val startPos = lookupPositionForOffset(rlsMetadata, offset)
     var remoteSegInputStream: InputStream = null
     try {
       // Search forward for the position of the last offset that is greater than or equal to the target offset
       remoteSegInputStream = remoteLogStorageManager.fetchLogSegmentData(rlsMetadata, startPos, Int.MaxValue)
       val remoteLogInputStream = new RemoteLogInputStream(remoteSegInputStream)
 
-      var batch:RecordBatch = null;
-      def fetchBatch() : RecordBatch = {
-        batch = remoteLogInputStream.nextBatch()
-        batch
-      }
-
-      // always look for the batch which has the desired offset and set the startPosition as that segments start
-      // location.
-      var lastBatchSize = 0
-
+      var firstBatch:RecordBatch = null
+      // Look for the batch which has the desired offset
       // we will always have a batch in that segment as it is a non-compacted topic. For compacted topics, we may need
       //to read from the subsequent segments if there is no batch available for the desired offset in the current
-      //segment. Tha means, desired offset is more than last offset of the current segment and immediate available
+      //segment. That means, desired offset is more than last offset of the current segment and immediate available
       //offset exists in the next segment which can be higher than the desired offset.
-      while (fetchBatch() != null && batch.lastOffset < offset) {
-        // sets the startPos as the starting of the next batch's start location.
-        lastBatchSize = batch.sizeInBytes()
-        startPos += lastBatchSize
+      while ((firstBatch = remoteLogInputStream.nextBatch()) != null && firstBatch.lastOffset < offset) {
```

Comment:
This throws a warning saying `comparing unrelated types Unit and null, it always returns false`. This is the reason why I introduced `fetchBatch()` method.

Reply:
Let us take case of batches b1=[0,50], b2=[51,100], b3=[101, 150] and b4= [150, 200]

For offsets
``` 
0  : b1
49 : b1
100: b2
200: b4 
```
This looks to be good.